packages/liveblocks-client/src/room.ts

```diff
@@ -730,69 +707,54 @@ export function makeStateMachine<
-    liveList: LiveList<TItem>,
-    callback: (liveList: LiveList<TItem>) => void
-  ): () => void;
-  function subscribe(
-    node: LiveStructure,
-    callback: (updates: StorageUpdate[]) => void,
-    options: { isDeep: true }
-  ): () => void;
-  function subscribe<TPresence extends JsonObject>(
-    type: "my-presence",
-    listener: MyPresenceCallback<TPresence>
-  ): () => void;
-  function subscribe<TPresence extends JsonObject>(
-    type: "others",
-    listener: OthersEventCallback<TPresence, TUserMeta>
-  ): () => void;
-  function subscribe(
-    type: "event",
-    listener: EventCallback<TEvent>
-  ): () => void;
-  function subscribe(type: "error", listener: ErrorCallback): () => void;
-  function subscribe(
-    type: "connection",
-    listener: ConnectionCallback
-  ): () => void;
-  function subscribe<K extends RoomEventName>(
-    firstParam: StorageCallback | K | LiveStructure,
-    listener?: RoomEventCallbackMap<TPresence, TUserMeta, TEvent>[K] | any,
+  // Generic storage callbacks
+  function subscribe(callback: StorageCallback): () => void; // prettier-ignore
+  // Storage callbacks filtered by Live structure
+  function subscribe<L extends LiveStructure>(liveStructure: L, callback: (node: L) => void): () => void; // prettier-ignore
+  function subscribe(node: LiveStructure, callback: StorageCallback, options: { isDeep: true }): () => void; // prettier-ignore
+  // Room event callbacks
+  function subscribe<E extends RoomEventName>(type: E, listener: RoomEventCallbackFor<E, TPresence, TUserMeta, TEvent>): () => void; // prettier-ignore
+
+  function subscribe<L extends LiveStructure, E extends RoomEventName>(
+    first: StorageCallback | L | E,
+    second?: ((node: L) => void) | StorageCallback | RoomEventCallback,
     options?: { isDeep: boolean }
   ): () => void {
-    if (isLiveNode(firstParam)) {
-      return subscribeToLiveStructure(firstParam, listener, options);
-    } else if (typeof firstParam === "function") {
-      return genericSubscribe(firstParam);
-    } else if (!isRoomEventName(firstParam)) {
-      throw new Error(`"${firstParam}" is not a valid event name`);
-    }
-
-    (
-      state.listeners[firstParam] as RoomEventCallbackMap<
-        TPresence,
-        TUserMeta,
-        TEvent
-      >[K][]
-    ).push(listener);
+    if (second === undefined || typeof first === "function") {
+      if (typeof first === "function") {
+        const storageCallback = first;
+        return genericSubscribe(storageCallback);
+      } else {
+        throw new Error("Please specify a listener callback");
+      }
+    }
+
+    if (isLiveNode(first)) {
+      const node = first;
+      if (options?.isDeep) {
+        const storageCallback = second as StorageCallback;
+        return subscribeToLiveStructureDeeply(node, storageCallback);
+      } else {
+        const nodeCallback = second as (node: L) => void;
+        return subscribeToLiveStructureShallowly(node, nodeCallback);
+      }
```

Comment:
By splitting up this helper into two versions, they're both much easier to type and implement. See their implementation to see how much simpler this code is to read and reason about now :)

Reply:
So much cleaner! 😮